Synchronize inserted, updated and deleted to Zoho CRM

rbergDrox opened the pull request:
Synchronisation issues #2 
Add arguments in the ZohoCopyDatabaseCommand to copy or synchronize.
ZohoDatabaseSyncZoho instance to push data to Zoho CRM.
Change the triggers to match with an unique identifier (insert or update) and use the zoho_id when we are deleting.
Unit Tests for ZohoDatabaseSyncZoho.

Commit:
Force run trigger

Commit:
Prepare insert, update, delete by Module

Commit:
Add the ZohoSync to Zoho CRM command

Commit:
Add the ZohoSync to Zoho CRM command:
Fix : 
method using to delete
passed variables in constructor

Commit:
Fix: Add uid field for each table to work with insertion

Commit:
Sync use getFlatFields as ZohoDatabaseCopier

Commit:
Fix :
Find uid to delete rows in local_update and local_delete

Commit:
Fix insertion method
Update id field

Commit:
Test insert to Zoho CRM

Commit:
remove return beans in insertion

Commit:
Perform test
Delete rows after insert, update, delete in local tables
Improve delete trigger to find id or uid (because the row is deleted in the module table)

Commit:
add sync command in ZohoCopyDatabaseCommand

Commit:
Copy and Sync method

rbergDrox committed:
Update README.md

Commit:
Fix no declared prefix
and non existed ZohoDaos

Commit:
Synchro : 
Update only all fields having changed
deal with date, datetime
Unit test to check values after updating

Commit:
Fix duplicate code by using ZohoDatabaseHelper::getTableName

Commit:
Fix :
unused code, unreachable statement

Commit:
Fix :
not existed parameters

moufmouf commented on README.md:
````diff
@@ -51,19 +54,38 @@ everything, you can use the second parameter of the `copy` method:
+$databaseSync->pushUpdatedRows($contactZohoDao);
+//pushDeletedRows method to delete the deleted rows to Zoho
+$databaseSync->pushDeletedRows($contactZohoDao);
+```
````
We should make a pass to rename objects / methods.

Maybe something like:

`ZohoDatabaseCopier::copy` => `ZohoDatabaseCopier::fetchFromZoho`

`ZohoDatabaseSyncZoho::xxx` => `ZohoDatabasePusher::pushToZoho` (the `pushToZoho` method calls in turn the pushInertedRows, pushUpdatedRows and pushDeletedRows). This makes a cleaner public interface to the class.

moufmouf commented on README.md:
```diff
@@ -51,19 +54,38 @@ everything, you can use the second parameter of the `copy` method:
+# Command to copy data
+$ console zoho:copy-db copy
+# Command to synchronize data
+$ console zoho:copy-db sync
```
Here too, we need a single command that synchronizes both ways.

Maybe:

```
console zoho:copy-db sync => both ways
console zoho:copy-db sync --fetch-only => copy
console zoho:copy-db sync --push-only => push
```

moufmouf commented on src/ZohoCopyDatabaseCommand.php:
```diff
@@ -27,36 +28,72 @@ class ZohoCopyDatabaseCommand extends Command
+                $this->syncDb($output);
+                break;
+            default:
+                throw new InvalidArgumentException('Named argument not found.');
```
Actually, text should be: 'Named argument must be one of 'copy' or 'sync'.' (but this should be changed anyway)

moufmouf commented on src/ZohoCopyDatabaseCommand.php:
```diff
@@ -27,36 +28,72 @@ class ZohoCopyDatabaseCommand extends Command
         try {
-            if ($this->lock) {
-                $this->lock->acquireLock();
+            if ($this->lockCopy) {
```
We have 2 different locks for the 2 actions. It means that technically, it is possible to trigger a copy and a sync at the same time. Couldn't it be dangerous? Maybe it is safer to put a same lock for both actions?

moufmouf commented on src/ZohoDatabaseCopier.php:
```diff
@@ -70,12 +69,12 @@ public function setLogger(LoggerInterface $logger)
             $this->localChangesTracker->createTrackingTables();
         }
-        $this->synchronizeDbModel($dao, $twoWaysSync);
+        $this->synchronizeDbModel($dao, $twoWaysSync, $forceCreateTrigger);
```
The `synchronizeDbModel` is in the ZohoDatabaseCopier class, but come to think about it, it is as much needed by the sync part than by the Zoho part. Maybe we should put it in its own class?

moufmouf commented on src/ZohoDatabaseSyncZoho.php:
```diff
@@ -0,0 +1,225 @@
+     * @param AbstractZohoDao $zohoDao
+     * @param string $localTable
+     */
+    public function pushDataToZoho(AbstractZohoDao $zohoDao, $localTable, $update = false){
```
This method should be private.
Also, the signature is weird. $localTable can be guessed from the $update parameter.
We should really drop this parameter and replace it with:

$localTable = $update ? 'local_update' : 'local_insert';

moufmouf commented on src/ZohoDatabaseSyncZoho.php:
```diff
@@ -0,0 +1,225 @@
+            $zohoDao->save($zohoBeans);
+            if(!$update){
+                foreach ($zohoBeans as $uid => $zohoBean) {
+                    $this->connection->update($tableName, [ 'id'=>$zohoBean->getZohoId(),'lastActivityTime'=> date("Y-m-d H:i:s") ], ['uid'=>$uid ]);
```
Why do we need to update lastActivityTime? This seems dangerous to me as any modification in Zoho before this lastActivityTime will not be fetched from Zoho. We should really not update it at all.

moufmouf commented on src/ZohoDatabaseSyncZoho.php:
```diff
@@ -0,0 +1,225 @@
+            }
+            $statementDelete = $this->connection->prepare('delete from '.$localTable.' where uid in ( :rowsDeleted)');
+            $statementDelete->execute([
+                'rowsDeleted' => implode(',', $rowsDeleted)
```
I'm pretty sure the `implode` will cause the query to fail if there is more than one deleted row.
Instead, you should rely on the PARAM_INT_ARRAY type of Doctrine. See: http://doctrine-orm.readthedocs.io/projects/doctrine-dbal/en/latest/reference/data-retrieval-and-manipulation.html#list-of-parameters-conversion

moufmouf commented on tests/ZohoDatabaseSyncZohoTest.php:
```diff
@@ -0,0 +1,117 @@
+        $resultDelete= $statementTestDelete->execute()->fetchAll();
+        $this->assertNotFalse($resultDelete);
+        $zohoZync->pushDeletedRows($contactZohoDao);
+        
```
We should add a test here to try to fetch the data from Zoho by ID and see that it throws an error.

moufmouf commented:
Also, use PHP-CS-FIXER

Commit:
Add new class to synchronize model + test
Improve ZohoDataPusher (  ex ZohoDatabaseSyncZoho)
Change method signature
Fix trigger update
Change command : zoho:sync ,unique locker file

Commit:
remove localTable param from pushdatatozoho

Commit:
Php-cs-fixer (applied)

Commit:
Fix: phpunit.xml.dist

moufmouf commented:
Closing to merge on 2.0 instead.
